packages/tailwindcss/src/plugin-api.ts

```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
         })
       }
     },
+    theme(path) {
+      let cssValue = readFromCss(designSystem.theme, path)
+
+      if (typeof cssValue !== 'object') {
+        return cssValue
+      }
+
+      let result = resolvedConfig.theme ?? {}
+
+      // 1. Convert `path` to an array of parts (segment)
+      // 2. Find the "obeject" that corresponds to the path
+      // 3. Deep merge the object with the resolved theme
+      let parts = segment(path, '.')
+
+      let obj = get(result, parts) ?? {}
+
+      if (typeof obj === 'object') {
+        deepMerge(obj, [cssValue], (a, b) => b)
+        set(result, parts, obj)
+      } else {
+        // This case happens when the CSS theme produces an object for a given
+        // keypath but the v3-style config does not. However we still want the
+        // CSS to "win" in this case so we overwrite the v3-style config value.
+        set(result, parts, cssValue)
+      }
+
+      for (let part of parts) {
+        result = result?.[part]
+      }
+
+      return result
+    },
   }
 }
 
 export function registerPlugins(plugins: Plugin[], designSystem: DesignSystem, ast: AstNode[]) {
-  let pluginApi = buildPluginApi(designSystem, ast)
+  let pluginObjects = []
 
   for (let plugin of plugins) {
     if ('__isOptionsFunction' in plugin) {
       // Happens with `plugin.withOptions()` when no options were passed:
       // e.g. `require("my-plugin")` instead of `require("my-plugin")(options)`
-      plugin().handler(pluginApi)
+      pluginObjects.push(plugin())
     } else if ('handler' in plugin) {
       // Happens with `plugin(…)`:
       // e.g. `require("my-plugin")`
       //
       // or with `plugin.withOptions()` when the user passed options:
       // e.g. `require("my-plugin")(options)`
-      plugin.handler(pluginApi)
+      pluginObjects.push(plugin)
     } else {
       // Just a plain function without using the plugin(…) API
-      plugin(pluginApi)
+      pluginObjects.push({ handler: plugin, config: {} as UserConfig })
+    }
+  }
+
+  // Now merge all the configs and make all that crap work
+  let resolvedConfig = resolveConfig([
+    createCompatabilityConfig(designSystem.theme),
+    ...pluginObjects.map(({ config }) => config ?? {}),
+  ])
+
+  let pluginApi = buildPluginApi(designSystem, ast, resolvedConfig)
```

Comment:
Can we do this conditionally based on whether the legacy APIs were used? I think ideally we don't pay any costs if you don't use any legacy stuff.

Maybe a first step is to only do this if we have `pluginObjects` at all?

Reply:
Yeah ideally we'd lazily create this only as necessary. I have some ideas but I think we need to consider its use by the CSS theme function before doing that.

Reply:
We'll handle this in @philipp-spiess's CSS theme fn PR as that will likely inform the design.